File: src/db/schema.ts
import {
  bigint,
  boolean,
  index,
  integer,
  pgTable,
  serial,
  text,
  timestamp,
} from 'drizzle-orm/pg-core'

export const users = pgTable('users', {
  telegramId: bigint('telegram_id', { mode: 'number' }).primaryKey(),
  name: text('name'),
  createdAt: timestamp('created_at', { withTimezone: true })
    .defaultNow()
    .notNull(),
})

export const tracks = pgTable(
  'tracks',
  {
    id: serial('id').primaryKey(),
    appleTrackId: text('apple_track_id').notNull().unique(),
    messageId: integer('message_id').notNull(),
    fileId: text('file_id').notNull(),
    fileUniqueId: text('file_unique_id').notNull(),
    title: text('title').notNull(),
    artist: text('artist').notNull(),
    album: text('album').notNull(),
    duration: integer('duration').notNull(),
    bitDepth: integer('bit_depth').notNull(),
    sampleRate: integer('sample_rate').notNull(),
    genre: text('genre').notNull(),
    releaseDate: text('release_date').notNull(),
    trackNumber: integer('track_number').notNull(),
    trackCount: integer('track_count').notNull(),
    createdAt: timestamp('created_at', { withTimezone: true })
      .defaultNow()
      .notNull(),
    updatedAt: timestamp('updated_at', { withTimezone: true })
      .defaultNow()
      .notNull(),
  },
  (table) => [
    index('tracks_title_idx').on(table.title),
    index('tracks_artist_idx').on(table.artist),
  ],
)

export const requests = pgTable(
  'requests',
  {
    id: serial('id').primaryKey(),
    telegramId: bigint('telegram_id', { mode: 'number' }).notNull(),
    chatId: bigint('chat_id', { mode: 'number' }).notNull(),
    appleTrackId: text('apple_track_id').notNull(),
    isCacheHit: boolean('is_cache_hit').notNull(),
    durationMs: integer('duration_ms'),
    status: text('status').notNull(),
    errorReason: text('error_reason'),
    createdAt: timestamp('created_at', { withTimezone: true })
      .defaultNow()
      .notNull(),
  },
  (table) => [
    index('requests_apple_track_id_idx').on(table.appleTrackId),
    index('requests_telegram_id_idx').on(table.telegramId),
    index('requests_created_at_idx').on(table.createdAt),
  ],
)

export type User = typeof users.$inferSelect
export type NewUser = typeof users.$inferInsert

export type Track = typeof tracks.$inferSelect
export type NewTrack = typeof tracks.$inferInsert

export type Request = typeof requests.$inferSelect
export type NewRequest = typeof requests.$inferInsert

File: crates/db/src/schema.rs
diesel::table! {
    users (telegram_id) {
        telegram_id -> BigInt,
        name -> Nullable<Text>,
        created_at -> Timestamptz,
    }
}

diesel::table! {
    tracks (id) {
        id -> Integer,
        provider -> Varchar,
        track_id -> Text,
        codec -> Varchar,
        message_id -> Integer,
        file_id -> Text,
        file_unique_id -> Text,
        title -> Text,
        artist -> Text,
        album -> Text,
        duration -> Integer,
        bit_depth -> Integer,
        sample_rate -> Integer,
        genre -> Text,
        release_date -> Text,
        track_number -> Integer,
        track_count -> Integer,
        isrc -> Nullable<Text>,
        created_at -> Timestamptz,
        updated_at -> Timestamptz,
    }
}

diesel::table! {
    requests (id) {
        id -> Integer,
        telegram_id -> BigInt,
        chat_id -> BigInt,
        provider -> Varchar,
        track_id -> Text,
        is_cache_hit -> Bool,
        duration_ms -> Nullable<Integer>,
        status -> Text,
        error_reason -> Nullable<Text>,
        created_at -> Timestamptz,
    }
}

diesel::table! {
    settings (id) {
        id -> SmallInt,
        data -> Jsonb,
        updated_at -> Timestamptz,
    }
}

diesel::table! {
    albums (id) {
        id -> Integer,
        provider -> Varchar,
        album_id -> Text,
        codec -> Varchar,
        part_index -> Integer,
        total_parts -> Integer,
        message_id -> Integer,
        file_id -> Text,
        file_unique_id -> Text,
        file_size -> BigInt,
        file_name -> Text,
        generation_hash -> Varchar,
        created_at -> Timestamptz,
    }
}

diesel::table! {
    user_sessions (id) {
        id -> Text,
        telegram_id -> BigInt,
        refresh_token_hash -> Text,
        device_name -> Nullable<Text>,
        platform -> Nullable<Text>,
        created_at -> Timestamptz,
        last_active_at -> Timestamptz,
        expires_at -> Timestamptz,
        revoked -> Bool,
    }
}

diesel::table! {
    one_time_auth_codes (code) {
        code -> Varchar,
        telegram_id -> BigInt,
        created_at -> Timestamptz,
        expires_at -> Timestamptz,
    }
}

diesel::table! {
    user_favorites (telegram_id, track_id) {
        telegram_id -> BigInt,
        track_id -> Integer,
        created_at -> Timestamptz,
    }
}

diesel::table! {
    user_playlists (id) {
        id -> Integer,
        telegram_id -> BigInt,
        name -> Text,
        created_at -> Timestamptz,
        updated_at -> Timestamptz,
    }
}

diesel::table! {
    user_playlist_tracks (playlist_id, track_id) {
        playlist_id -> Integer,
        track_id -> Integer,
        position -> Integer,
        created_at -> Timestamptz,
    }
}

diesel::table! {
    tg_worker_sessions (bot_token_hash) {
        bot_token_hash -> Varchar,
        session_data -> Text,
        created_at -> Timestamptz,
        updated_at -> Timestamptz,
    }
}

diesel::table! {
    user_integrations (telegram_id, provider) {
        telegram_id -> BigInt,
        provider -> Varchar,
        username -> Text,
        encrypted_session_key -> Text,
        created_at -> Timestamptz,
        updated_at -> Timestamptz,
    }
}

diesel::allow_tables_to_appear_in_same_query!(
    users,
    tracks,
    requests,
    settings,
    albums,
    user_sessions,
    one_time_auth_codes,
    user_favorites,
    user_playlists,
    user_playlist_tracks,
    tg_worker_sessions,
    user_integrations
);
